File: detect.py
import os
import json
import requests
import re
from http.server import BaseHTTPRequestHandler


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        # Pre-flight check for browsers
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', 'https://hexascope.lovable.app')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
        self.end_headers()

    def do_POST(self):
        # 1. SETUP HEADERS
        allowed_origin = "https://hexascope.lovable.app"

        # 2. READ INPUT DATA
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length)
        body = json.loads(post_data)
        base64_image = body.get('image')

        # 3. CALL OPENAI
        api_key = os.environ.get("OPENAI_API_KEY")
        openai_url = "https://api.openai.com/v1/chat/completions"

        payload = {
            "model": "gpt-4.1-mini",
            "max_tokens": 600,
            "messages": [{
                "role": "user",
                "content": [
                    {"type": "text", "text": "You are an expert entomologist... (Your full prompt here)"},
                    {"type": "image_url", "image_url": {"url": f"data:image/jpeg;base64,{base64_image}"}}
                ]
            }]
        }

        try:
            response = requests.post(
                openai_url,
                headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
                json=payload
            )

            res_json = response.json()
            raw_content = res_json['choices'][0]['message']['content']
            cleaned = re.sub(r'```json|```', '', raw_content).strip()

            # 4. SEND RESPONSE
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', allowed_origin)
            self.end_headers()
            self.wfile.write(cleaned.encode('utf-8'))

        except Exception as e:
            self.send_response(500)
            self.end_headers()
            self.wfile.write(json.dumps({"error": str(e)}).encode('utf-8'))

File: api/detect.js
export default async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') return res.status(200).end();
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' });

  const { image } = req.body;
  if (!image) return res.status(200).json({ insect_name: 'No Image', bio: 'No image provided.' });

  try {
    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${process.env.OPENAI_API_KEY}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        model: 'gpt-4o-mini',
        max_tokens: 1000,
        messages: [{
          role: 'user',
          content: [
            { type: 'text', text: `You are an expert entomologist. Identify the insect in this image. Accept ALL real-world photos including phone camera shots. Do NOT reject images for quality. If an insect is held by hands or gloves focus only on the insect. Return ONLY raw JSON no markdown no backticks with these exact fields: insect_name, biological_name, genus, species, classification (object with kingdom phylum class order family genus species), confidence (decimal 0-1 based on actual certainty not always 0.95), danger_level (low/medium/high/unknown), school_risk_level (low/medium/high/unknown), diet, lifespan, habitat, population, economic_importance, bio. If no insect set insect_name to No Insect Detected. If unclear set insect_name to Unknown Insect.` },
            { type: 'image_url', image_url: { url: image } }
          ]
        }]
      })
    });

    const data = await response.json();

    if (data.error) {
      return res.status(200).json({ insect_name: 'Connection Issue', bio: 'AI is busy. Please try again.' });
    }

    const raw = data.choices[0].message.content;
    const cleaned = raw.replace(/```json|```/g, '').trim();
    const result = JSON.parse(cleaned);
    return res.status(200).json(result);

  } catch (e) {
    return res.status(200).json({ insect_name: 'Error', bio: `Something went wrong: ${e.message}` });
  }
}
